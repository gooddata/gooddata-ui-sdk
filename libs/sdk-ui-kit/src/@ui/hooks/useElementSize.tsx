// (C) 2024-2025 GoodData Corporation
import { useLayoutEffect, useState, useRef } from "react";

/**
 * @internal
 */
export function useElementSize() {
    const ref = useRef<HTMLElement | null>(null);
    const [height, setHeight] = useState<number>(0);
    const [width, setWidth] = useState<number>(0);

    useLayoutEffect(() => {
        const resizeObserver = new ResizeObserver((entries) => {
            const [entry] = entries;
            if (entry) {
                setHeight(entry.contentRect.height);
                setWidth(entry.contentRect.width);
            }
        });

        if (ref.current) {
            resizeObserver.observe(ref.current);
        }

        return () => {
            if (resizeObserver) {
                resizeObserver.disconnect();
                setHeight(0);
                setWidth(0);
            }
        };
    }, []);

    return {
        ref,
        height,
        width,
    };
}
