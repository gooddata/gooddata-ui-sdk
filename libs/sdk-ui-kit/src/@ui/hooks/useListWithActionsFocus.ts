// (C) 2025 GoodData Corporation

import { useCallback, useEffect, useRef } from "react";

import { createContextStore } from "@gooddata/sdk-ui";

import { useIdPrefixed } from "../../utils/useId.js";

/**
 * @internal
 */
export function useFocusWithinContainer(idToFocus?: string | null) {
    const containerRef = useRef<HTMLElement>(null);

    useEffect(() => {
        const container = containerRef.current;
        const escapedIdToFocus = idToFocus ? CSS.escape(idToFocus) : null;
        const elementToFocus = idToFocus ? container?.querySelector(`#${escapedIdToFocus}`) : null;

        if (!elementToFocus) {
            return;
        }

        if (!container?.contains(document.activeElement) && container !== document.activeElement) {
            return;
        }

        (elementToFocus as HTMLElement).focus();
    }, [idToFocus]);

    return { containerRef };
}

/**
 * @internal
 */
export const useListWithActionsFocusStoreValue = <T>(getIdFromItem: (item: T) => string) => {
    const containerId = useIdPrefixed("focus-store");

    const makeId = useCallback(
        ({ item, action }: { item: T; action: string }) =>
            CSS.escape(`${containerId}-${getIdFromItem(item)}-${action}`.replace(/\s/g, "_")),
        [containerId, getIdFromItem],
    );

    return { makeId, containerId };
};

/**
 * @internal
 */
export const ListWithActionsFocusStore =
    createContextStore<ReturnType<typeof useListWithActionsFocusStoreValue>>("ListWithActionsFocusStore");
