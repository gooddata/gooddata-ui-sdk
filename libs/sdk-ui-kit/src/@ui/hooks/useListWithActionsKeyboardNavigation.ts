// (C) 2025 GoodData Corporation

import React from "react";
import { useAutoupdateRef } from "@gooddata/sdk-ui";
import { makeLinearKeyboardNavigation, makeMenuKeyboardNavigation } from "../@utils/keyboardNavigation.js";

/**
 * @internal
 */
export const SELECT_ITEM_ACTION = "selectItem";

interface IKeyboardNavigationDeps<Item, Action extends string> {
    items: Item[];
    focusedAction: Action | typeof SELECT_ITEM_ACTION;
    focusedItemAdditionalActions: Action[];
    focusedItem: Item | undefined;
    actionHandlers: {
        [key in Action | typeof SELECT_ITEM_ACTION]: (
            item: Item,
            e?: React.KeyboardEvent,
        ) => (() => void) | undefined;
    };
    setFocusedIndex: React.Dispatch<React.SetStateAction<number | undefined>>;
    setFocusedAction: React.Dispatch<React.SetStateAction<Action | typeof SELECT_ITEM_ACTION>>;
}

/**
 * @internal
 */
export function useListWithActionsKeyboardNavigation<Item, Action extends string>({
    items,
    actionHandlers,
    getItemAdditionalActions,
    isNestedList = false,
    focusedIndex: focusedIndexProp,
}: {
    items: Item[];
    actionHandlers: {
        [key in Action | typeof SELECT_ITEM_ACTION]: (
            item: Item,
            e?: React.KeyboardEvent,
        ) => (() => void) | undefined;
    };
    getItemAdditionalActions: (item: Item) => Action[];
    isNestedList?: boolean;
    focusedIndex?: number;
}) {
    const [focusedIndex, setFocusedIndex] = React.useState<number | undefined>(focusedIndexProp ?? 0);
    const focusedItem = focusedIndex === undefined ? undefined : items[focusedIndex];

    const [focusedAction, setFocusedAction] = React.useState<Action | typeof SELECT_ITEM_ACTION>(
        SELECT_ITEM_ACTION,
    );

    React.useEffect(() => {
        if (focusedIndexProp !== undefined) {
            setFocusedIndex(focusedIndexProp);
        }
    }, [focusedIndexProp]);

    const focusedItemAdditionalActions = React.useMemo(
        () => (focusedItem ? getItemAdditionalActions(focusedItem) : []),
        [focusedItem, getItemAdditionalActions],
    );

    // If the items change and we're suddenly focusing on a nonexistent item
    if (focusedIndex && focusedIndex > items.length - 1) {
        setFocusedIndex(items.length - 1);
    }
    // If the item actions change and we're suddenly focusing on a nonexistent action
    if (focusedAction !== SELECT_ITEM_ACTION && !focusedItemAdditionalActions.includes(focusedAction)) {
        setFocusedAction(focusedItemAdditionalActions[0] ?? SELECT_ITEM_ACTION);
    }

    const keyboardNavigationDepsRef = useAutoupdateRef<IKeyboardNavigationDeps<Item, Action>>({
        items,
        focusedAction,
        setFocusedIndex,
        setFocusedAction,
        focusedItemAdditionalActions,
        focusedItem,
        actionHandlers,
    });

    // There are two separate modes of control,
    // one switches between items and one controls picking actions for a specific item
    const selectionMode = focusedAction === SELECT_ITEM_ACTION ? "item" : "additionalAction";

    const handleKeyboardNavigation = React.useMemo(() => {
        return selectionMode === "item"
            ? // Selecting items
              makeItemSelectionNavigation(keyboardNavigationDepsRef)
            : // Selecting actions
              makeActionSelectionNavigation(keyboardNavigationDepsRef, isNestedList);
    }, [keyboardNavigationDepsRef, selectionMode, isNestedList]);

    const handleBlur = React.useCallback<React.FocusEventHandler>(() => {
        setFocusedAction(SELECT_ITEM_ACTION);
    }, []);

    return {
        onKeyboardNavigation: handleKeyboardNavigation,
        onBlur: handleBlur,
        focusedAction,
        focusedItem,
        setFocusedAction,
    };
}

function makeItemSelectionNavigation<Item, Action extends string>(
    depsRef: React.MutableRefObject<IKeyboardNavigationDeps<Item, Action>>,
) {
    return makeMenuKeyboardNavigation({
        onFocusPrevious: () => {
            const { items, setFocusedIndex } = depsRef.current;

            setFocusedIndex((currentIndex) =>
                currentIndex === undefined || currentIndex === 0 ? items.length - 1 : currentIndex - 1,
            );
        },
        onFocusNext: () => {
            const { items, setFocusedIndex } = depsRef.current;

            setFocusedIndex((currentIndex) =>
                currentIndex === undefined || currentIndex === items.length - 1 ? 0 : currentIndex + 1,
            );
        },
        onFocusFirst: () => {
            const { setFocusedIndex } = depsRef.current;

            setFocusedIndex(0);
        },
        onFocusLast: () => {
            const { items, setFocusedIndex } = depsRef.current;

            setFocusedIndex(items.length - 1);
        },
        onSelect: (e) => {
            const { actionHandlers, focusedItem } = depsRef.current;

            if (!focusedItem) {
                return;
            }

            actionHandlers.selectItem(focusedItem, e)?.();
        },
        onEnterLevel: () => {
            const { focusedItemAdditionalActions, setFocusedAction } = depsRef.current;

            if (focusedItemAdditionalActions.length === 0) {
                return;
            }

            setFocusedAction(focusedItemAdditionalActions[0]);
        },
    });
}

function makeActionSelectionNavigation<Item, Action extends string>(
    depsRef: React.MutableRefObject<IKeyboardNavigationDeps<Item, Action>>,
    isNestedList: boolean,
) {
    return makeLinearKeyboardNavigation({
        onFocusPrevious: (e) => {
            const { items, focusedItemAdditionalActions, setFocusedIndex, setFocusedAction } =
                depsRef.current;
            if (isNestedList && e.key === "ArrowUp") {
                setFocusedIndex((currentIndex) =>
                    currentIndex === undefined || currentIndex === 0 ? items.length - 1 : currentIndex - 1,
                );
                setFocusedAction(SELECT_ITEM_ACTION);

                return;
            }

            setFocusedAction((currentAction) => {
                if (currentAction === SELECT_ITEM_ACTION) {
                    return currentAction;
                }

                return (
                    focusedItemAdditionalActions[focusedItemAdditionalActions.indexOf(currentAction) - 1] ??
                    focusedItemAdditionalActions.at(-1) ??
                    SELECT_ITEM_ACTION
                );
            });
        },
        onFocusNext: (e) => {
            const { items, focusedItemAdditionalActions, setFocusedIndex, setFocusedAction } =
                depsRef.current;

            if (isNestedList && e.key === "ArrowDown") {
                setFocusedIndex((currentIndex) =>
                    currentIndex === undefined || currentIndex === items.length - 1 ? 0 : currentIndex + 1,
                );
                setFocusedAction(SELECT_ITEM_ACTION);

                return;
            }

            setFocusedAction((currentAction) => {
                if (currentAction === SELECT_ITEM_ACTION) {
                    return currentAction;
                }

                return (
                    focusedItemAdditionalActions[focusedItemAdditionalActions.indexOf(currentAction) + 1] ??
                    focusedItemAdditionalActions[0] ??
                    SELECT_ITEM_ACTION
                );
            });
        },
        onFocusFirst: () => {
            const { focusedItemAdditionalActions, setFocusedAction } = depsRef.current;

            setFocusedAction(focusedItemAdditionalActions[0] ?? SELECT_ITEM_ACTION);
        },
        onFocusLast: () => {
            const { focusedItemAdditionalActions, setFocusedAction } = depsRef.current;

            setFocusedAction(focusedItemAdditionalActions.at(-1) ?? SELECT_ITEM_ACTION);
        },
        onSelect: () => {
            const { actionHandlers, focusedAction, focusedItem } = depsRef.current;

            if (!focusedItem) {
                return;
            }

            actionHandlers[focusedAction](focusedItem)?.();
        },
        onClose: () => {
            const { setFocusedAction } = depsRef.current;

            setFocusedAction(SELECT_ITEM_ACTION);
        },
    });
}
