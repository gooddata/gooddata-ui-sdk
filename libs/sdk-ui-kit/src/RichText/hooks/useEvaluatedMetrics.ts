// (C) 2022-2025 GoodData Corporation

import {
    DataPoint,
    useBackend,
    useCancelablePromise,
    useExecutionDataView,
    useWorkspace,
} from "@gooddata/sdk-ui";
import {
    IFilter,
    IMeasure,
    IResultHeader,
    isMeasureDefinition,
    newMeasure,
    ObjRef,
} from "@gooddata/sdk-model";

import { ReferenceMap } from "../helpers/references.js";

export type EvaluatedMetric = {
    def: IMeasure;
    header: IResultHeader;
    ref: ObjRef;
    data: DataPoint;
};

export function useEvaluatedMetrics(references: ReferenceMap, filters: IFilter[], enabled: boolean) {
    const workspace = useWorkspace();
    const backend = useBackend();
    const metrics = getMeasures(references);

    const {
        status: executionStatus,
        result: executionResult,
        error: executionError,
    } = useExecutionDataView({
        execution:
            enabled && metrics.length > 0
                ? backend.workspace(workspace).execution().forItems(metrics, filters)
                : null,
    });

    const {
        status: loadStatus,
        result: loadResult,
        error: loadError,
    } = useCancelablePromise(
        {
            promise: async () => {
                if (!executionResult) {
                    return [];
                }

                const headerItems = executionResult.meta().allHeaders();
                const dataSeries = executionResult.data().series().toArray();
                const definition = executionResult.definition;
                const items: Array<EvaluatedMetric> = [];

                headerItems[0][0].forEach((header, i) => {
                    const def = definition.measures[i];
                    const data = dataSeries[i].dataPoints()[0];

                    if (isMeasureDefinition(def.measure.definition)) {
                        items.push({
                            def,
                            data,
                            header,
                            ref: def.measure.definition.measureDefinition.item,
                        });
                    }
                });
                return items;
            },
        },
        [executionResult],
    );

    return {
        loading: executionStatus === "loading" || loadStatus === "loading" || loadStatus === "pending",
        error: executionError || loadError,
        result: loadResult,
    };
}

function getMeasures(references: ReferenceMap) {
    const metrics = Object.values(references)
        .map(({ type, ref }) => (type === "measure" ? ref : null))
        .filter(Boolean);

    return metrics.map((ref, i) => {
        return newMeasure(ref, (m) => {
            m.localId(`m_${i}`);
            return m;
        });
    });
}
