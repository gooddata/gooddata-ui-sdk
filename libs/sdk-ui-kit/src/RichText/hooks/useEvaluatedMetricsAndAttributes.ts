// (C) 2022-2025 GoodData Corporation

import {
    DataPoint,
    useBackend,
    useCancelablePromise,
    useExecutionDataView,
    useWorkspace,
} from "@gooddata/sdk-ui";
import {
    IExecutionConfig,
    IFilter,
    IMeasure,
    IResultHeader,
    isMeasureDefinition,
    newAttribute,
    newMeasure,
    ObjRef,
} from "@gooddata/sdk-model";

import { ReferenceMap } from "../helpers/references.js";

export type EvaluatedMetric = {
    def: IMeasure;
    header: IResultHeader;
    ref: ObjRef;
    data: DataPoint;
    count: number;
    format?: string;
};

export function useEvaluatedMetricsAndAttributes(
    references: ReferenceMap,
    filters: IFilter[],
    config: IExecutionConfig & { enabled: boolean; isFiltersLoading?: boolean },
) {
    const workspace = useWorkspace();
    const backend = useBackend();
    const metrics = getMeasures(references);
    const { metrics: labels, countMap } = getLabels(references);

    const items = [...metrics, ...labels];
    const { enabled, isFiltersLoading, ...execConfig } = config;

    const {
        status: executionStatus,
        result: executionResult,
        error: executionError,
    } = useExecutionDataView(
        {
            execution:
                enabled && items.length > 0 && !isFiltersLoading
                    ? backend
                          .workspace(workspace)
                          .execution()
                          .forItems(items, filters)
                          .withExecConfig(execConfig)
                    : null,
        },
        [execConfig.timestamp, execConfig.dataSamplingPercentage, isFiltersLoading],
    );

    const {
        status: loadStatus,
        result: loadResult,
        error: loadError,
    } = useCancelablePromise(
        {
            promise: async () => {
                if (!executionResult) {
                    return [];
                }

                const headerItems = executionResult.meta().allHeaders();
                const dataSeries = executionResult.data().series().toArray();
                const definition = executionResult.definition;
                const items: Array<EvaluatedMetric> = [];

                const skippedLocalIds = Object.values(countMap);
                const valueLocalIds = Object.keys(countMap);

                const relatedHeaders = headerItems[0][0];
                relatedHeaders.forEach((header, i) => {
                    const data = dataSeries[i].dataPoints()[0];
                    const def = data.seriesDesc.measureDefinition;
                    const descriptor = data.seriesDesc.measureDescriptor;

                    // Skip count measures, we want merge it with value measures
                    if (skippedLocalIds.includes(def.measure.localIdentifier)) {
                        return;
                    }

                    // Normal measure
                    if (
                        isMeasureDefinition(def.measure.definition) &&
                        !valueLocalIds.includes(def.measure.localIdentifier)
                    ) {
                        items.push({
                            def,
                            data,
                            header,
                            count: 1,
                            format: descriptor.measureHeaderItem.format,
                            ref: def.measure.definition.measureDefinition.item,
                        });
                    }

                    // Attribute measure with count as next
                    if (
                        isMeasureDefinition(def.measure.definition) &&
                        valueLocalIds.includes(def.measure.localIdentifier)
                    ) {
                        const countLocalId = countMap[def.measure.localIdentifier];
                        const countIndex = definition.measures.findIndex(
                            (d) => d.measure.localIdentifier === countLocalId,
                        );
                        const countData = dataSeries[countIndex]?.dataPoints()[0] ?? undefined;

                        items.push({
                            def,
                            data,
                            header,
                            format: descriptor.measureHeaderItem.format,
                            count: parseInt((countData?.rawValue ?? "0").toString()),
                            ref: def.measure.definition.measureDefinition.item,
                        });
                    }
                });
                return items;
            },
        },
        [executionResult],
    );

    return {
        loading:
            isFiltersLoading ||
            executionStatus === "loading" ||
            loadStatus === "loading" ||
            loadStatus === "pending",
        error: executionError || loadError,
        result: loadResult,
    };
}

function getMeasures(references: ReferenceMap) {
    const metrics = Object.values(references)
        .map(({ type, ref }) => (type === "measure" ? ref : null))
        .filter(Boolean);

    return metrics.map((ref, i) => {
        return newMeasure(ref, (m) => {
            m.localId(`m_${i}`);
            return m;
        });
    });
}

function getLabels(references: ReferenceMap) {
    const attributes = Object.values(references)
        .map(({ type, ref }) => (type === "displayForm" || type === "attribute" ? ref : null))
        .filter(Boolean);

    const items = attributes.map((ref, i) => {
        return newAttribute(ref, (m) => {
            m.localId(`m_${i}`);
            return m;
        });
    });

    const countMap: Record<string, string> = {};
    const metrics = items.reduce((acc: IMeasure[], attr, i) => {
        const valueLocalId = `m_max_${i}`;
        const countLocalId = `m_count_${i}`;

        countMap[valueLocalId] = countLocalId;
        return [
            ...acc,
            newMeasure(attr.attribute.displayForm, (m) => {
                m.localId(valueLocalId);
                m.aggregation("max");
                return m;
            }),
            newMeasure(attr.attribute.displayForm, (m) => {
                m.localId(countLocalId);
                m.aggregation("count");
                return m;
            }),
        ];
    }, []);

    return {
        countMap,
        metrics,
    };
}
