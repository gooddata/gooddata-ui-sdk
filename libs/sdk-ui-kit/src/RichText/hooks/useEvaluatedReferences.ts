// (C) 2025 GoodData Corporation
import { useMemo } from "react";
import { IExecutionConfig, IFilter } from "@gooddata/sdk-model";

import { collectReferences } from "../helpers/references.js";

import { useEvaluatedMetricsAndAttributes } from "./useEvaluatedMetricsAndAttributes.js";

export function useEvaluatedReferences(
    value: string,
    filters: IFilter[],
    config: IExecutionConfig & { enabled: boolean; isFiltersLoading?: boolean },
) {
    const isEmptyValue = useMemo(() => !value?.replace(/\s/g, ""), [value]);
    const references = useMemo(() => collectReferences(value), [value]);

    const {
        loading,
        result: metrics,
        error: metricsError,
    } = useEvaluatedMetricsAndAttributes(references, filters, config);

    return {
        isEmptyValue,
        loading,
        metrics,
        error: metricsError,
    };
}
