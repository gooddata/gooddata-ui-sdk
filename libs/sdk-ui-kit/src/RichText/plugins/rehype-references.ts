// (C) 2022-2025 GoodData Corporation

import { areObjRefsEqual, IdentifierRef, ISeparators } from "@gooddata/sdk-model";
import { ClientFormatterFacade } from "@gooddata/number-formatter";
import { IntlShape } from "react-intl";
import { Parent } from "unist";
import { Root } from "mdast";
import cx from "classnames";

import { EvaluatedMetric } from "../hooks/useEvaluatedMetricsAndAttributes.js";
import { createReference } from "../helpers/references.js";

import { HtmlNode, REFERENCE_REGEX_MATCH, REFERENCE_REGEX_SPLIT, TextNode } from "./types.js";

export function rehypeReferences(intl: IntlShape, metrics?: EvaluatedMetric[], separators?: ISeparators) {
    return function () {
        return function (tree: Root) {
            iterateTree(tree as HtmlNode, {
                onTextNodeReference: (text, ref) => {
                    const metric = metrics?.find((m) => areObjRefsEqual(m.ref, ref));
                    const { metricDef } = createMetricValue(intl, text, metric, separators);
                    return [metricDef];
                },
                onTextRawReference: (ref) => {
                    const metric = metrics?.find((m) => areObjRefsEqual(m.ref, ref));
                    const { value, formattedValue } = createMetricValue(intl, null, metric);
                    return value !== null ? formattedValue : "";
                },
            });
            return tree;
        };
    };
}

function iterateTree(
    node: HtmlNode,
    callbacks: {
        onTextNodeReference: (text: TextNode, ref: IdentifierRef, id: string) => Parent[];
        onTextRawReference: (ref: IdentifierRef, id: string) => string;
    },
): Parent[] {
    //Text type
    if (node.type === "text") {
        const res = iterateReferenceMatch((node as TextNode).value, (ref, id) => {
            return callbacks.onTextNodeReference(node as TextNode, ref, id);
        });
        return res.length ? res : [node];
    }
    //Image type
    if (node.type === "element" && node.tagName === "img") {
        if (node.properties.alt) {
            node.properties.alt = iterateReferenceRawTextMatch(String(node.properties.alt), (ref, id) => {
                return callbacks.onTextRawReference(ref, id);
            });
        }
        if (node.properties.title) {
            node.properties.title = iterateReferenceRawTextMatch(String(node.properties.title), (ref, id) => {
                return callbacks.onTextRawReference(ref, id);
            });
        }
        if (node.properties.src) {
            const src = decodeURI(node.properties.src);
            const update = iterateReferenceRawTextMatch(String(src), (ref, id) => {
                return callbacks.onTextRawReference(ref, id);
            });
            node.properties.src = encodeURI(update);
        }
    }
    //Anchor type
    if (node.type === "element" && node.tagName === "a") {
        if (node.properties.title) {
            node.properties.title = iterateReferenceRawTextMatch(String(node.properties.title), (ref, id) => {
                return callbacks.onTextRawReference(ref, id);
            });
        }
        if (node.properties.href) {
            const url = decodeURI(node.properties.href);
            const updated = iterateReferenceRawTextMatch(String(url), (ref, id) => {
                return callbacks.onTextRawReference(ref, id);
            });
            node.properties.href = encodeURI(updated);
        }
    }
    if (node.children) {
        // has children
        node.children = node.children.reduce((acc, child) => {
            return [...acc, ...iterateTree(child as HtmlNode, callbacks)];
        }, []);
        return [node];
    }

    // no children
    return [node];
}

function iterateReferenceRawTextMatch(
    value: string,
    onMatch: (ref: IdentifierRef, id: string) => string,
): string {
    const items = value.split(REFERENCE_REGEX_SPLIT);
    return items
        .map((item) => {
            const match = REFERENCE_REGEX_MATCH.exec(item);
            if (match) {
                const { ref } = createReference(match);
                if (ref) {
                    return onMatch(ref, match[2]);
                }
            }
            return item;
        })
        .join("");
}

function iterateReferenceMatch<T>(value: string, onMatch: (ref: IdentifierRef, id: string) => T[]): T[] {
    const items: T[] = [];
    REFERENCE_REGEX_MATCH.lastIndex = -1;
    let match = REFERENCE_REGEX_MATCH.exec(value);
    while (match) {
        const { ref } = createReference(match);
        if (ref) {
            items.push(...onMatch(ref, match[2]));
        }
        match = REFERENCE_REGEX_MATCH.exec(value);
    }
    return items;
}

function createMetricValue(
    intl: IntlShape,
    text: TextNode | null,
    metric: EvaluatedMetric,
    separators?: ISeparators,
) {
    const { formattedValue, value, isEmpty, isMultiple, styles } = getValue(metric, intl, separators);

    const metricDef = {
        type: "element",
        tagName: "span",
        properties: {
            className: cx({
                "gd-rich-text-metric-error": !metric,
                "gd-rich-text-metric-empty": metric && isEmpty,
                "gd-rich-text-metric-multiple": metric && isMultiple,
                "gd-rich-text-metric-value": metric && !isEmpty && !isMultiple,
            }),
            style: styles,
        },
        position: text?.position ?? undefined,
        children: [{ type: "text", value: formattedValue }],
    };

    return {
        metricDef,
        formattedValue,
        value: isEmpty || isMultiple ? null : value,
    };
}

function getValue(metric: EvaluatedMetric, intl: IntlShape, separators?: ISeparators) {
    const value = metric ? metric.data.rawValue : null;
    const isMultiple = metric?.count && metric.count > 1;
    const isEmpty = value === null || value === undefined || metric?.count === 0;

    let styles: Partial<CSSStyleDeclaration> = {};

    if (!metric) {
        return {
            formattedValue: `(${intl.formatMessage({ id: "richText.no_fetch" })})`,
            value,
            styles,
            isMultiple,
            isEmpty,
        };
    }
    if (isEmpty) {
        return {
            formattedValue: `(${intl.formatMessage({ id: "richText.no_data" })})`,
            value,
            styles,
            isMultiple,
            isEmpty,
        };
    }
    if (isMultiple) {
        return {
            formattedValue: `(${intl.formatMessage({ id: "richText.multiple_data" })})`,
            value,
            styles,
            isMultiple,
            isEmpty,
        };
    }

    const { formattedValue, colors } = createNumberJsFormatter(metric, separators);
    styles = {
        ...(colors?.color ? { color: colors.color } : {}),
        ...(colors?.backgroundColor ? { backgroundColor: colors.backgroundColor } : {}),
    };

    // NOTE: If the results value is empty, we should display the empty value message
    if (formattedValue == "") {
        return {
            formattedValue: `(${intl.formatMessage({ id: "empty_value" })})`,
            isEmpty: true,
            isMultiple,
            styles,
            value,
        };
    }

    // Default case
    return {
        formattedValue,
        isEmpty,
        isMultiple,
        styles,
        value,
    };
}

function createNumberJsFormatter(metric: EvaluatedMetric, separators?: ISeparators) {
    const value = metric ? metric.data.rawValue : null;

    if (!metric?.data.formatable) {
        return {
            formattedValue: value as string,
            colors: {
                color: undefined,
                backgroundColor: undefined,
            },
        };
    }

    const valueToFormat = ClientFormatterFacade.convertValue(value);
    const { formattedValue, colors } = ClientFormatterFacade.formatValue(
        valueToFormat,
        metric.format,
        separators,
    );

    return {
        formattedValue,
        colors,
    };
}
