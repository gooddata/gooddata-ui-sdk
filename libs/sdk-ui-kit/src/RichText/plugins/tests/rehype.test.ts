// (C) 2025 GoodData Corporation

import { describe, expect, it } from "vitest";
import { Root } from "mdast";

import { createIntlMock, DataPoint } from "@gooddata/sdk-ui";
import { rehypeReferences } from "../rehype-references";
import { EvaluatedMetric } from "../../hooks/useEvaluatedMetricsAndAttributes";

describe("testing rehype plugin to extract references", () => {
    const intl = createIntlMock();
    const metrics: EvaluatedMetric[] = [
        em("metric_1", "Metric 1", 100, true),
        em("metric_2", "Metric 2", 200, true),
    ];
    const metrics1: EvaluatedMetric[] = [
        em("metric_1", "Metric 1", 100, false),
        em("metric_2", "Metric 2", 200, false),
    ];

    const htmlTreeSimpleText = {
        type: "root",
        children: [
            {
                type: "element",
                tagName: "p",
                properties: {},
                children: [
                    {
                        type: "text",
                        value: "Hello, there are some ",
                    },
                    {
                        type: "text",
                        value: "{metric/metric_1}",
                    },
                    {
                        type: "text",
                        value: " and ",
                    },
                    {
                        type: "text",
                        value: "{metric/metric_2}",
                    },
                    {
                        type: "text",
                        value: " references.",
                    },
                ],
            },
        ],
    };
    const htmlTreeSimpleText2 = {
        type: "root",
        children: [
            {
                type: "element",
                tagName: "p",
                properties: {},
                children: [
                    {
                        type: "text",
                        value: "Hello, there are some ",
                    },
                    {
                        type: "text",
                        value: "{metric/metric_1}",
                    },
                    {
                        type: "text",
                        value: " and ",
                    },
                    {
                        type: "text",
                        value: "{metric/metric_2}",
                    },
                    {
                        type: "text",
                        value: " references.",
                    },
                ],
            },
        ],
    };

    const htmlTreeLinkText = {
        type: "root",
        children: [
            {
                type: "element",
                tagName: "p",
                properties: {},
                children: [
                    {
                        type: "text",
                        value: "Hello, there are some ",
                    },
                    {
                        type: "element",
                        tagName: "a",
                        properties: {
                            href: "https://www.{metric/metric_1}.com",
                            title: "Link to {metric/metric_1}",
                        },
                        children: [
                            {
                                type: "text",
                                value: "{metric/metric_1}",
                            },
                        ],
                    },
                    {
                        type: "text",
                        value: " references.",
                    },
                ],
            },
        ],
    };

    const htmlTreeImageText = {
        type: "root",
        children: [
            {
                type: "element",
                tagName: "p",
                properties: {},
                children: [
                    {
                        type: "text",
                        value: "Hello, there is image ",
                    },
                    {
                        type: "element",
                        tagName: "img",
                        properties: {
                            alt: "This is {metric/metric_1} alt",
                            title: "Link to {metric/metric_1}",
                            src: "https://www.{metric/metric_1}.com",
                        },
                    },
                    {
                        type: "text",
                        value: " references.",
                    },
                ],
            },
        ],
    };

    it("replace references ids for real formatted values in text", () => {
        const walk = rehypeReferences(intl, metrics)();
        const updated = walk(htmlTreeSimpleText as Root);

        expect(updated).toMatchSnapshot();
    });

    it("replace references ids for real formatted values in links", () => {
        const walk = rehypeReferences(intl, metrics)();
        const updated = walk(htmlTreeLinkText as Root);

        expect(updated).toMatchSnapshot();
    });

    it("replace references ids for real formatted values in image", () => {
        const walk = rehypeReferences(intl, metrics)();
        const updated = walk(htmlTreeImageText as Root);

        expect(updated).toMatchSnapshot();
    });

    it("replace references ids for not formatted values in text", () => {
        const walk = rehypeReferences(intl, metrics1)();
        const updated = walk(htmlTreeSimpleText2 as Root);

        expect(updated).toMatchSnapshot();
    });
});

function em(id: string, title: string, value: number, formatable: boolean): EvaluatedMetric {
    return {
        ref: {
            type: "measure",
            identifier: id,
        },
        def: {
            measure: {
                title,
                localIdentifier: id,
                definition: {
                    inlineDefinition: {
                        maql: "SELECT SUM(*)",
                    },
                },
            },
        },
        header: {
            measureHeaderItem: {
                name: title,
                order: 0,
            },
        },
        data: {
            formatable,
            coordinates: [],
            rawValue: value,
        } as DataPoint,
        count: 1,
    };
}
