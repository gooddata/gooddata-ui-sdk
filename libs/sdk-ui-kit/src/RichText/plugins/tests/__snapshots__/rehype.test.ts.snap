// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`testing rehype plugin to extract references > replace references ids for not formatted values in text 1`] = `
{
  "children": [
    {
      "children": [
        {
          "type": "text",
          "value": "Hello, there are some ",
        },
        {
          "children": [
            {
              "type": "text",
              "value": 100,
            },
          ],
          "position": undefined,
          "properties": {
            "className": "gd-rich-text-metric-value",
            "style": {},
          },
          "tagName": "span",
          "type": "element",
        },
        {
          "type": "text",
          "value": " and ",
        },
        {
          "children": [
            {
              "type": "text",
              "value": 200,
            },
          ],
          "position": undefined,
          "properties": {
            "className": "gd-rich-text-metric-value",
            "style": {},
          },
          "tagName": "span",
          "type": "element",
        },
        {
          "type": "text",
          "value": " references.",
        },
      ],
      "properties": {},
      "tagName": "p",
      "type": "element",
    },
  ],
  "type": "root",
}
`;

exports[`testing rehype plugin to extract references > replace references ids for real formatted values in image 1`] = `
{
  "children": [
    {
      "children": [
        {
          "type": "text",
          "value": "Hello, there is image ",
        },
        {
          "properties": {
            "alt": "This is 100.00 alt",
            "src": "https://www.100.00.com",
            "title": "Link to 100.00",
          },
          "tagName": "img",
          "type": "element",
        },
        {
          "type": "text",
          "value": " references.",
        },
      ],
      "properties": {},
      "tagName": "p",
      "type": "element",
    },
  ],
  "type": "root",
}
`;

exports[`testing rehype plugin to extract references > replace references ids for real formatted values in links 1`] = `
{
  "children": [
    {
      "children": [
        {
          "type": "text",
          "value": "Hello, there are some ",
        },
        {
          "children": [
            {
              "children": [
                {
                  "type": "text",
                  "value": "100.00",
                },
              ],
              "position": undefined,
              "properties": {
                "className": "gd-rich-text-metric-value",
                "style": {},
              },
              "tagName": "span",
              "type": "element",
            },
          ],
          "properties": {
            "href": "https://www.100.00.com",
            "title": "Link to 100.00",
          },
          "tagName": "a",
          "type": "element",
        },
        {
          "type": "text",
          "value": " references.",
        },
      ],
      "properties": {},
      "tagName": "p",
      "type": "element",
    },
  ],
  "type": "root",
}
`;

exports[`testing rehype plugin to extract references > replace references ids for real formatted values in text 1`] = `
{
  "children": [
    {
      "children": [
        {
          "type": "text",
          "value": "Hello, there are some ",
        },
        {
          "children": [
            {
              "type": "text",
              "value": "100.00",
            },
          ],
          "position": undefined,
          "properties": {
            "className": "gd-rich-text-metric-value",
            "style": {},
          },
          "tagName": "span",
          "type": "element",
        },
        {
          "type": "text",
          "value": " and ",
        },
        {
          "children": [
            {
              "type": "text",
              "value": "200.00",
            },
          ],
          "position": undefined,
          "properties": {
            "className": "gd-rich-text-metric-value",
            "style": {},
          },
          "tagName": "span",
          "type": "element",
        },
        {
          "type": "text",
          "value": " references.",
        },
      ],
      "properties": {},
      "tagName": "p",
      "type": "element",
    },
  ],
  "type": "root",
}
`;
